src/main/scala/temple/DSL/Semantics/package.scala

```diff
@@ -0,0 +1,383 @@
+      registerKeyword(key, key, argType, constructor)
+
+    /** Perform parsing by looking up the relevant function and
+      * [[scala.collection.IterableOnceOps#foldRight(java.lang.Object, scala.Function2)]] supports */
+    final def apply(metaKey: String, args: Args)(implicit context: BlockContext): T =
+      matchers.get(metaKey).map(_(args)) getOrElse {
+        fail(s"No valid metadata $metaKey in $context")
+      }
+  }
+
+  /** A parser of Metadata items that can occur in all the blocks */
+  private val parseGlobalMetadata = new MetadataParser[GlobalMetadata] {
+    registerKeyword("language", TokenArgType, Language)
+  }
+
+  /** A parser of Metadata items that can occur in project and service blocks */
+  private val parseProjectAndServiceMetadata = new MetadataParser[ProjectAndServiceMetadata] {
+    inherit from parseGlobalMetadata
+    registerKeyword("database", TokenArgType, Database)
+  }
+
+  /** A parser of Metadata items that can occur in service blocks */
+  private val parseServiceMetadata = new MetadataParser[ServiceMetadata] {
+    inherit from parseProjectAndServiceMetadata
+    registerKeyword("auth", "login", TokenArgType, ServiceAuth)
+    registerKeyword("uses", "services", ListArgType(TokenArgType), Uses)
+  }
+
+  /** A parser of Metadata items that can occur in project blocks */
+  private val parseProjectMetadata = new MetadataParser[ProjectMetadata] {
+    inherit from parseProjectAndServiceMetadata
+    registerKeyword("provider", TokenArgType, Provider)
+  }
+
+  /** A parser of Metadata items that can occur in target blocks */
+  private val parseTargetMetadata = new MetadataParser[TargetMetadata] {
+    inherit from parseGlobalMetadata
+    registerKeyword("auth", "services", ListArgType(TokenArgType), TargetAuth)
+  }
+
+  /**
+    * Parse a service block from a list of entries into the distinct attributes, metadatas and structs
+    * @param entries The list of entries in the block from the AST
+    * @param context The location in the AST, used for error messages
+    * @return A semantic representation of a [[temple.DSL.Semantics.ServiceBlock]]
+    */
+  private def parseServiceBlock(entries: List[Entry])(implicit context: BlockContext): ServiceBlock = {
+    val attributes = mutable.LinkedHashMap[String, AttributeType]()
+    val metadatas  = mutable.ListBuffer[ServiceMetadata]()
+    val structs    = mutable.LinkedHashMap[String, StructBlock]()
+    entries.foreach {
+      case Entry.Attribute(key, dataType, annotations) =>
+        attributes.safeInsert(key -> parseAttributeType(dataType, annotations)(KeyName(key)))
+      case Entry.Metadata(metaKey, args) => metadatas += parseServiceMetadata(metaKey, args)
+      case DSLRootItem(key, tag, entries) =>
+        tag match {
+          case "struct" => structs.safeInsert(key -> parseStructBlock(entries)(BlockContext(key, tag, context)))
+          case tag      => fail(s"Unknown block type $tag for $key in $context")
+        }
```

Comment:
Should this be flattened? I guess so, despite the additional repetition of `DSLRootItem(key, _, entries)`